File: modules/platforms/cpp/common/src/common/bits.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <algorithm>
#include <cassert>

#include "ignite/common/bits.h"

namespace ignite
{
    namespace common
    {
        namespace bits
        {
            int32_t NumberOfTrailingZerosI32(int32_t i)
            {
                int32_t y;

                if (i == 0) return 32;

                int32_t n = 31;

                y = i << 16;

                if (y != 0) {
                    n = n - 16;
                    i = y;
                }

                y = i << 8; 

                if (y != 0) {
                    n = n - 8;
                    i = y;
                }

                y = i << 4;

                if (y != 0) {
                    n = n - 4;
                    i = y;
                }

                y = i << 2;

                if (y != 0) {
                    n = n - 2;
                    i = y;
                }

                return n - static_cast<int32_t>(static_cast<uint32_t>(i << 1) >> 31);
            }

            int32_t NumberOfLeadingZerosI32(int32_t i)
            {
                return NumberOfLeadingZerosU32(static_cast<uint32_t>(i));
            }

            int32_t NumberOfLeadingZerosU32(uint32_t i)
            {
                if (i == 0)
                    return 32;

                int32_t n = 1;

                if (i >> 16 == 0) {
                    n += 16;
                    i <<= 16;
                }

                if (i >> 24 == 0) {
                    n += 8;
                    i <<= 8;
                }

                if (i >> 28 == 0) {
                    n += 4;
                    i <<= 4;
                }

                if (i >> 30 == 0) {
                    n += 2;
                    i <<= 2;
                }

                return n - static_cast<int32_t>(i >> 31);
            }

            int32_t NumberOfLeadingZerosI64(int64_t i)
            {
                return NumberOfLeadingZerosU64(static_cast<uint64_t>(i));
            }

            int32_t NumberOfLeadingZerosU64(uint64_t i)
            {
                if (i == 0)
                    return 64;

                int32_t n = 1;

                uint32_t x = static_cast<uint32_t>(i >> 32);

                if (x == 0) {
                    n += 32;
                    x = static_cast<uint32_t>(i);
                }

                if (x >> 16 == 0) {
                    n += 16;
                    x <<= 16;
                }

                if (x >> 24 == 0) {
                    n += 8;
                    x <<= 8;
                }

                if (x >> 28 == 0) {
                    n += 4;
                    x <<= 4;
                }

                if (x >> 30 == 0) {
                    n += 2;
                    x <<= 2;
                }

                n -= x >> 31;

                return n;
            }

            int32_t BitCountI32(int32_t i)
            {
                uint32_t ui = static_cast<uint32_t>(i);

                ui -= (ui >> 1) & 0x55555555;
                ui = (ui & 0x33333333) + ((ui >> 2) & 0x33333333);
                ui = (ui + (ui >> 4)) & 0x0f0f0f0f;
                ui += ui >> 8;
                ui += ui >> 16;

                return static_cast<int32_t>(ui & 0x3f);
            }

            int32_t BitLengthI32(int32_t i)
            {
                return 32 - NumberOfLeadingZerosI32(i);
            }

            int32_t BitLengthU32(uint32_t i)
            {
                return 32 - NumberOfLeadingZerosU32(i);
            }

            int32_t GetCapasityForSize(int32_t size)
            {
                assert(size > 0);

                if (size <= 8)
                    return 8;

                int32_t bl = BitLengthI32(size);

                if (bl > 30)
                    return INT32_MAX;

                int32_t res = 1 << bl;

                return size > res ? res << 1 : res;
            }

            int32_t DigitLength(uint64_t x)
            {
                // See http://graphics.stanford.edu/~seander/bithacks.html
                // for the details on the algorithm.

                if (x < 10)
                    return 1;

                int32_t r = ((64 - NumberOfLeadingZerosU64(x) + 1) * 1233) >> 12;

                assert(r <= UINT64_MAX_PRECISION);

                return (r == UINT64_MAX_PRECISION || x < TenPowerU64(r)) ? r : r + 1;
            }

            uint64_t TenPowerU64(int32_t n)
            {
                static const uint64_t TEN_POWERS_TABLE[UINT64_MAX_PRECISION] = {
                    1U,                     // 0  / 10^0
                    10U,                    // 1  / 10^1
                    100U,                   // 2  / 10^2
                    1000U,                  // 3  / 10^3
                    10000U,                 // 4  / 10^4
                    100000U,                // 5  / 10^5
                    1000000U,               // 6  / 10^6
                    10000000U,              // 7  / 10^7
                    100000000U,             // 8  / 10^8
                    1000000000U,            // 9  / 10^9
                    10000000000U,           // 10 / 10^10
                    100000000000U,          // 11 / 10^11
                    1000000000000U,         // 12 / 10^12
                    10000000000000U,        // 13 / 10^13
                    100000000000000U,       // 14 / 10^14
                    1000000000000000U,      // 15 / 10^15
                    10000000000000000U,     // 16 / 10^16
                    100000000000000000U,    // 17 / 10^17
                    1000000000000000000U,   // 18 / 10^18
                    10000000000000000000U   // 19 / 10^19
                };

                assert(n >= 0 && n < UINT64_MAX_PRECISION);

                return TEN_POWERS_TABLE[n];
            }
        }
    }
}

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/Bits.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.network.serialization.marshal;

/**
 * Code for packing/unpacking primitive values to/from a byte array at specific offsets.
 */
class Bits {
    static boolean getBoolean(byte[] b, int off) {
        return b[off] != 0;
    }

    static char getChar(byte[] b, int off) {
        return (char) ((b[off + 1] & 0xFF)
                + (b[off] << 8));
    }

    static short getShort(byte[] b, int off) {
        return (short) ((b[off + 1] & 0xFF)
                + (b[off] << 8));
    }

    static int getInt(byte[] b, int off) {
        return ((b[off + 3] & 0xFF))
                + ((b[off + 2] & 0xFF) << 8)
                + ((b[off + 1] & 0xFF) << 16)
                + ((b[off]) << 24);
    }

    static float getFloat(byte[] b, int off) {
        return Float.intBitsToFloat(getInt(b, off));
    }

    static long getLong(byte[] b, int off) {
        return ((b[off + 7] & 0xFFL))
                + ((b[off + 6] & 0xFFL) << 8)
                + ((b[off + 5] & 0xFFL) << 16)
                + ((b[off + 4] & 0xFFL) << 24)
                + ((b[off + 3] & 0xFFL) << 32)
                + ((b[off + 2] & 0xFFL) << 40)
                + ((b[off + 1] & 0xFFL) << 48)
                + (((long) b[off]) << 56);
    }

    static double getDouble(byte[] b, int off) {
        return Double.longBitsToDouble(getLong(b, off));
    }

    static void putBoolean(byte[] b, int off, boolean val) {
        b[off] = (byte) (val ? 1 : 0);
    }

    static void putChar(byte[] b, int off, char val) {
        b[off + 1] = (byte) (val);
        b[off] = (byte) (val >>> 8);
    }

    static void putShort(byte[] b, int off, short val) {
        b[off + 1] = (byte) (val);
        b[off] = (byte) (val >>> 8);
    }

    static void putInt(byte[] b, int off, int val) {
        b[off + 3] = (byte) (val);
        b[off + 2] = (byte) (val >>> 8);
        b[off + 1] = (byte) (val >>> 16);
        b[off] = (byte) (val >>> 24);
    }

    static void putFloat(byte[] b, int off, float val) {
        putInt(b, off, Float.floatToIntBits(val));
    }

    static void putLong(byte[] b, int off, long val) {
        b[off + 7] = (byte) (val);
        b[off + 6] = (byte) (val >>> 8);
        b[off + 5] = (byte) (val >>> 16);
        b[off + 4] = (byte) (val >>> 24);
        b[off + 3] = (byte) (val >>> 32);
        b[off + 2] = (byte) (val >>> 40);
        b[off + 1] = (byte) (val >>> 48);
        b[off] = (byte) (val >>> 56);
    }

    static void putDouble(byte[] b, int off, double val) {
        putLong(b, off, Double.doubleToLongBits(val));
    }
}
